third_party/maya/lib/usdMaya/usdWriteJob.cpp

```diff
@@ -221,45 +212,44 @@ bool usdWriteJob::beginJob(const std::string &iFileName,
             continue;
         }
 
-        MayaPrimWriterPtr primWriter = nullptr;
-        if (!createPrimWriter(curDagPath, &primWriter) &&
-                curDagPath.length() > 0) {
+        if (!needToTraverse(curDagPath) &&
+            curDagPath.length() > 0) {
             // This dagPath and all of its children should be pruned.
             itDag.prune();
-            continue;
-        }
-
-        if (primWriter) {
-            mMayaPrimWriterList.push_back(primWriter);
-
-            // Write out data (non-animated/default values).
-            if (UsdPrim usdPrim = primWriter->write(UsdTimeCode::Default())) {
-                MDagPath dag = primWriter->getDagPath();
-                mDagPathToUsdPathMap[dag] = usdPrim.GetPath();
-
-                // If we are merging transforms and the object derives from
-                // MayaTransformWriter but isn't actually a transform node, we
-                // need to add its parent.
-                if (mArgs.mergeTransformAndShape) {
-                    MayaTransformWriterPtr xformWriter =
-                            boost::dynamic_pointer_cast<MayaTransformWriter>(
-                                    primWriter);
-                    if (xformWriter) {
-                        MDagPath xformDag = xformWriter->getTransformDagPath();
-                        mDagPathToUsdPathMap[xformDag] = usdPrim.GetPath();
+        } else {
+            MayaPrimWriterPtr primWriter = createPrimWriter(curDagPath);
+
+            if (primWriter) {
+                mMayaPrimWriterList.push_back(primWriter);
+
+                primWriter->write(UsdTimeCode::Default());
+
+                // Write out data (non-animated/default values).
+                if (const auto& usdPrim = primWriter->getPrim()) {
+                    MDagPath dag = primWriter->getDagPath();
+                    mDagPathToUsdPathMap[dag] = usdPrim.GetPath();
+
+                    // If we are merging transforms and the object derives from
+                    // MayaTransformWriter but isn't actually a transform node, we
+                    // need to add its parent.
+                    if (mArgs.mergeTransformAndShape) {
+                        MayaTransformWriterPtr xformWriter =
+                            std::dynamic_pointer_cast<MayaTransformWriter>(primWriter);
+                        if (xformWriter) {
+                            MDagPath xformDag = xformWriter->getTransformDagPath();
+                            mDagPathToUsdPathMap[xformDag] = usdPrim.GetPath();
+                        }
                     }
                 }
 
-                mModelKindWriter.OnWritePrim(usdPrim, primWriter);
```

Comment:
No, there were various rewrites while working on this change, and I forgot to add this back to the code. (it was present somewhere else for a while) Added it back.